storagemarket/impl/providerstates/provider_states_test.go

```diff
@@ -401,6 +403,57 @@ func TestEnsureProviderFunds(t *testing.T) {
 	}
 }
 
+func TestRestartDataTransfer(t *testing.T) {
+	channelId := datatransfer.ChannelID{Initiator: peer.ID("1"), Responder: peer.ID("2")}
+	ctx := context.Background()
+	eventProcessor, err := fsm.NewEventProcessor(storagemarket.MinerDeal{}, "State", providerstates.ProviderEvents)
+	require.NoError(t, err)
+	runRestartDataTransfer := makeExecutor(ctx, eventProcessor, providerstates.RestartDataTransfer, storagemarket.StorageDealProviderTransferRestart)
+	tests := map[string]struct {
+		nodeParams        nodeParams
+		dealParams        dealParams
+		environmentParams environmentParams
+		fileStoreParams   tut.TestFileStoreParams
+		pieceStoreParams  tut.TestPieceStoreParams
+		dealInspector     func(t *testing.T, deal storagemarket.MinerDeal, env *fakeEnvironment)
+	}{
+		"succeeds": {
+			dealParams: dealParams{
+				TransferChannelId: &channelId,
+			},
+			dealInspector: func(t *testing.T, deal storagemarket.MinerDeal, env *fakeEnvironment) {
+				require.Eventually(t, func() bool {
+					return len(env.restartDataTransferCalls) == 1
+				}, 5*time.Second, 200*time.Millisecond)
+				require.Equal(t, channelId, env.restartDataTransferCalls[0].chId)
+				tut.AssertDealState(t, storagemarket.StorageDealProviderTransferRestart, deal.State)
+			},
+		},
+		// TODO FIXME
+		"RestartDataTransfer errors": {
+			dealParams: dealParams{
+				TransferChannelId: &channelId,
+			},
+			environmentParams: environmentParams{
+				RestartDataTransferError: xerrors.New("some error"),
+			},
+			dealInspector: func(t *testing.T, deal storagemarket.MinerDeal, env *fakeEnvironment) {
+				require.Eventually(t, func() bool {
+					fmt.Printf("\n deal state is %s", storagemarket.DealStates[deal.State])
```

Comment:
This isn't going to work unfortunately.

The problem is really hard to see but the state of the deal only updates when you call ReplayEvents on the FSM TestContext (see line 1228 of this file)

Sorry I hadn't thought about any of these tests being asynchronous. 

Fixing this is non-trivial. Given that, I'm ok if you comment out the `RestartDataTransfer errors` test for now (leave the code there till we can figure something out)

Reply:
You'll see if you look at the CI that this test is failing -- my guess is if it works on your local machine it's just cause your local machine is faster